File: Python/Bronze/2019-12/WhereAmI.py
#
#  USACO 2019 December - Bronze - Problem 2 - Where Am I?
#
import sys

sys.stdin = open('whereami.in', 'r')
sys.stdout = open('whereami.out', 'w')

N = int(input())
S = input()

for K in range(1, N+1):
	substr_set = set()
	unique = True

	for i in range(N-K+1):
		substr = S[i:i+K]
		if substr in substr_set:
			unique = False
			break
		substr_set.add(substr)
	
	if unique:
		print(K)
		break

File: CPlusPlus/Bronze/2019-12/WhereAmI.cpp
/**
 *  USACO 2019 December - Bronze - Problem 2 - Where Am I?
 */

#include <bits/stdc++.h>

using namespace std;

int main()
{
    freopen("whereami.in", "r", stdin);
    freopen("whereami.out", "w", stdout);

    int N;
    cin >> N;

    string str;
    cin >> str;

    for (int K = 1; K <= N; K++)
    {
        unordered_set<string> substrSet;
        bool unique = true;

        for (int i = 0; i <= N-K; i++)
        {
            string substr = str.substr(i, K);
            if (substrSet.count(substr)) {
                unique = false;
                break;
            }
            substrSet.insert(substr);
        }

        if (unique) {
            cout << K << endl;
            break;
        }
    }

    return 0;
}